cypress/support/step_definitions/dictionary/addBulkConceptsToDictionary/addBulkConceptsToDictionary.ts

```diff
@@ -0,0 +1,141 @@
+/// <reference types="cypress" />
+/// <reference types="../../../" />
+import { 
+  After, 
+  Before, 
+  Given, 
+  Then, 
+  When } from "cypress-cucumber-preprocessor/steps";
+import { getDictionaryId, getUser } from '../../../utils';
+
+Given("the user is on the view dictionary concepts page", () => {
+  cy.visit(`/users/${getUser()}/collections/${getDictionaryId()}/concepts/`);
+  cy.url().should(
+    "contain",
+    `/users/${getUser()}/collections/${getDictionaryId()}/concepts/`
+  );
+});
+
+Given('the user is on the "Add concepts in bulk from CIEL" page', () => {
+  cy.visit(
+    `/users/${getUser()}/collections/${getDictionaryId()}/add/?fromSource=CIEL/`
+  );
+  
+  cy.url().should(
+    "contain",
+    `/users/${getUser()}/collections/${getDictionaryId()}/add/?fromSource=CIEL/`
+  );
+});
+
+Given('CIEL concept "1000" is already in the dictionary', () => {
+  cy.getConcept(`/users/${getUser()}/collections/${getDictionaryId()}/concepts/1000`).should("exist");
+});
+
+When('the user clicks the "Add concepts" floating action button', () => {
+  cy.findByTitle("Add concepts").click();
+});
+
+When('the user clicks the "Add concepts" MuiButton-label button', () => {
+  cy.findByText("Add concepts").click();
+});
+
+When('the user selects "Add bulk concepts"', () => {
+  cy.visit(
+    `/users/${getUser()}/collections/${getDictionaryId()}/add/?fromSource=CIEL/`
+  );
+});
+
+When(
+  /the user enters concept Id "(1000|1001|1002)"/, (conceptId) => {
+    cy.get("textarea").type(conceptId)
+  }
+); 
+
+When('the user re-enters concept Id "1000"', (conceptId) => {
+    cy.get("textarea").type(conceptId)
+  }
+); 
+
+When('the user navigates to the "Progress notifications" page', () => {
+  cy.visit(
+    `/actions/`
+  );
+});
+
+Then('the user selects "Import existing concept"', () => {
+   cy.findByText("Import existing concept").click();
+});
+
+Then('the user should be on the "Add concepts in bulk from CIEL" page', () => {
+   cy.url().should("contain", `/users/${getUser()}/collections/${getDictionaryId()}/add/?fromSource=CIEL/`);
+});
+
+Then(
+  /the concept Id "1000" should be added to the dictionary/, conceptId => {
```

Comment:
What's the value of `conceptId` here? It would be helpful to look at the examples where we using similar arguments, maybe in other tests, of perhaps at line 79 where this works.

Reply:
To be a bit more explicit: the value of `conceptId` here is... I actually have no idea. Usually, this works because we use a regular expression with a capture group, e.g.

`/the concept Id "(1000|1001|1002)" should be in the dictionary/`

Is a regular expression that will match any of:

```
the concept Id "1000" should be in the dictionary
the concept Id "1001" should be in the dictionary
the concept Id "1002" should be in the dictionary
```

The `|` here expresses the alternating patterns and the parentheses indicate a capture group.

You can try this out in the console of your browser. For example,

```js
/my test (1000|2000|3000)/.exec("my test 1000");
```

Will give you an array like this:

```js
[ "my test 1000", "1000" ]
```

Whereas this:

```js
/my test (1000|2000|3000)/.exec("1000");
```

results in `null` (because the "my test" text is missing).

Behind the scenes, the Cypress Cucumber library will take the matching array as shown above and convert it to arguments, so that with this:

```js
Given(/the concept Id "(1000|1001|1002)" should be in the dictionary/, conceptId => {});
```

and the line in a feature file `the concept Id "1000" should be in the dictionary`, the value of the `conceptId` variable will be `1000`.

But where we don't have a regex or a capture group... I honestly don't know what the value will be. I'd expect it to be `undefined`.

That's what the failures you're pointing to about, though, is that you're taking the variable with an undefined value and passing it into a URL to construct it, which returns a 404 error because from the backend's perspective you're requesting a URL that doesn't exist.

Reply:
yes @ibacher that would be the issue.
@jwnasambu , since you are trying to pass a value to the function, we could use this:
```suggestion
  /the concept Id {string} should be added to the dictionary/, conceptId => {
```

Reply:
see:
https://github.com/TheBrainFamily/cypress-cucumber-preprocessor#reusable-step-definitions

Reply:
@jayasanka-sack Shouldn't it be `"{string}"`, i.e. we want the value to be `1000`, not `"1000"`, but if the step in the feature file is written `the concept Id "1000" should be added to the dictionary`, `{string}` should match the `"1000"` part.

Reply:
@ibacher and @jayasanka-sack  I have tried out the proposed changes in my terminal but still running into different errors.